crates/backend/src/codegen.rs

```diff
@@ -586,6 +587,98 @@ impl ToTokens for ast::ImportType {
     }
 }
 
+impl ToTokens for ast::ImportEnum {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let vis = &self.vis;
+        let name = &self.name;
+        let name_string = &self.name.to_string();
+        let variants = &self.variants;
+        let variant_strings = &self.variant_values;
+
+        let mut current_idx: usize = 0;
+        let variant_indexes: Vec<Literal> = variants
+            .iter()
+            .map(|_| {
+                let this_index = current_idx;
+                current_idx += 1;
+                Literal::usize_unsuffixed(this_index)
+            })
+            .collect();
+
+        // Borrow variant_indexes because we need to use it multiple times inside the quote! macro
+        let variant_indexes_ref = &variant_indexes;
+
+        // A vector of EnumName::VariantName tokens for this enum
+        let variant_paths: Vec<TokenStream> = self
+            .variants
+            .iter()
+            .map(|v| quote!(#name::#v).into_token_stream())
+            .collect();
+
+        // Borrow variant_paths because we need to use it multiple times inside the quote! macro
+        let variant_paths_ref = &variant_paths;
+
+        (quote! {
+            #[allow(bad_style)]
+            #[derive(Copy, Clone, Debug)]
+            #vis enum #name {
+                #(#variants = #variant_indexes_ref,)*
+            }
+
+            impl ::wasm_bindgen::describe::WasmDescribe for #name {
+                fn describe() {
+                    ::wasm_bindgen::JsValue::describe()
+                }
+            }
+
+            impl ::wasm_bindgen::convert::IntoWasmAbi for #name {
+                type Abi = <::wasm_bindgen::JsValue as
+                    ::wasm_bindgen::convert::IntoWasmAbi>::Abi;
+
+                fn into_abi(self, extra: &mut ::wasm_bindgen::convert::Stack) -> Self::Abi {
+                    match self {
+                        #(#variant_paths_ref => ::wasm_bindgen::JsValue::from_str(#variant_strings).into_abi(extra),)*
+                    }
+                }
+            }
+
+            impl ::wasm_bindgen::convert::FromWasmAbi for #name {
+                type Abi = <::wasm_bindgen::JsValue as
+                    ::wasm_bindgen::convert::FromWasmAbi>::Abi;
+
+                unsafe fn from_abi(
+                    js: Self::Abi,
+                    extra: &mut ::wasm_bindgen::convert::Stack,
+                ) -> Self {
+                    #name::from(::wasm_bindgen::JsValue::from_abi(js, extra))
+                }
+            }
+
+            impl From<::wasm_bindgen::JsValue> for #name {
+                fn from(obj: ::wasm_bindgen::JsValue) -> #name {
+                    let obj_str = match obj.as_string() {
+                        Some(string_value) => string_value,
+                        None => panic!("Can't convert a non-string into {}", #name_string),
```

Comment:
I agree that when we receive a non-string value or non-enum string value as an argument to a function that expects this enum, we should panic.

However, I don't think we should implement `From<JsValue>` for these and have it panic. This is too much of a footgun for folks who are just reading docs and see the impl and don't know what it does under the hood.

Instead, I think we should make a normal, non-trait `pub fn from_js_value(JsValue) -> Option<Self>` method, and the `FromWasmAbi` trait impl can `unwrap` the result. This gives us the same behavior when sending them to and from the ABI, but without the footgun.

Reply:
I think panicking in `FromWasmAbi` is something we should avoid so that we can recover in cases where JS functions return invalid string values.
### Example ###
```webidl
enum Foo { "bar", "baz" }
Foo getFoo();
```

```js
getFoo() {
    return "quux"
}
```

```rust
// Panics
match getFoo() {
    Foo::Bar => ...,
    Foo::Baz => ...,
}
```

### Possible Solutions ###
* We could use this [idiom from libstd](https://github.com/rust-lang/rust/blob/master/src/libstd/io/error.rs#L190) and return `__Invalid` instead of panicking
```rust
enum Foo {
    Bar,
    Baz,
    #[doc(hidden)]
    __Invalid,
}
```
* We could just expose a catch-all variant
```rust
enum Foo {
    Bar,
    Baz,
    InvalidValue(String),
}
```
* Or we could make WebIDL functions which return enums return `Result`s.

Reply:
+1. I definitely wouldn't want my web app to panic if a browser gets updated with an updated version of a standard, and I haven't gotten around to upgrading my copy of `web-sys`.

The main downside to including the invalid value in the actual enum is that the enum no longer can be copy. I would be inclined to either just include an `__Invalid` variant, or have WebIDL functions returning enums to instead return `Result`s, with an error type that contains the invalid string.

Reply:
I like the `Result` solution the best, it seems like it's the most convenient for the end user. The only downside is that functions that throw will end up returning `Result<Result<Enum, String>, ...>`, but there might be a nice way to flatten that out.

Reply:
Another advantage of having non exhaustive `enum`s is that we won't have to do breaking version bumps when updating WebIDLs.

Reply:
I am working on adding support for partial and mixin interfaces right now. As part of those changes, I am doing a two pass process that would make it trivial to implement the `Result` return type solution. (This is in part to fix a bug in the current crate where interface objects are being passed by value, rather than by reference, as per the spec.)

I should have time today to finish up a first draft, and submit a PR. It is based off of this branch, so I would return to the non-exhaustive `enum` problem in a follow-up PR (especially if you want to go down the `Result` path).

Reply:
@ohanar Sounds good. I'll make `from_js_value` visible and resolve the merge conflict, then I'm good with this being merged.

@fitzgen That's a good point, definitely a huge + for non-exhaustive over `Result`s.